allennlp/models/model.py

```diff
@@ -253,7 +253,13 @@ def _load(cls,
 
         # Load vocabulary from file
         vocab_dir = os.path.join(serialization_dir, 'vocabulary')
-        vocab = Vocabulary.from_files(vocab_dir)
+        vocab_type = Vocabulary
+        if "vocabulary" in config:
+            vocab_params = config["vocabulary"]
+            vocab_type = vocab_params.pop("type", None)
+            if vocab_type is not None:
+                vocab_type = Vocabulary.by_name(vocab_type)
```

Comment:
What about what we do here: https://github.com/allenai/allennlp/blob/580dc8b0e2c6491d4d75b54c3b15b34b462e0c67/allennlp/nn/initializers.py#L64-L65

Reply:
probably that's a better way to do it, the tricky part is that `Vocabulary` is both the base class and the default implementation, so right now it's not even registered as a type of Vocabulary (and you couldn't use the decorator for that, you'd have to do a little hack after the class is defined)

Reply:
Having `Vocabulary.register(Vocabulary, 'default_vocabulary')`, or something, right after the class definition, doesn't seem like that big of a deal.  And it simplifies this logic here enough that I'd vote for it.

Reply:
yeah, this occurred to me after I made the comment.

@WrRan does this make sense? basically you'll need to

1. in the `Vocabulary` class, set a default implementation (maybe just call it `'default'`), like we do here: https://github.com/allenai/allennlp/blob/580dc8b0e2c6491d4d75b54c3b15b34b462e0c67/allennlp/data/tokenizers/tokenizer.py#L24

2. after the class is defined, call `Vocabulary.register('default')(Vocabulary)` like mattg suggested (we can't use the decorator like we usually do because `Vocabulary` hasn't been defined yet when you'd want to use it)

3. then you'd have to do something like

```
vocab_params = config.pop('vocabulary', Params({}))
vocab_choice = vocab_params.pop_choice('type', Vocabulary.list_available())
vocab = Vocabulary.by_name(vocab_choice).from_files(vocab_dir)
```

I think that's right, you might have to tweak it a tiny bit.

Reply:
Minor point - the function is only curried when it's used as a decorator, so I'm pretty sure it really should just be `Vocabulary.register(Vocabulary, 'default')`.

Reply:
it's curried either way, that's how it's written. the decorator syntax just hides the second function call from you.

```
In [1]: from allennlp.common import Registrable

In [2]: class A(Registrable): pass

In [3]: A.register('default', A)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-32a13d43632d> in <module>()
----> 1 A.register('default', A)

TypeError: register() takes 2 positional arguments but 3 were given

In [4]: A.register('default')(A)
Out[4]: __main__.A

In [5]: A.by_name('default')
Out[5]: __main__.A
```

Reply:
Oh, sorry, my bad.  I saw that the function had two arguments, and I misunderstood them.  You're right.